src/tools/cephfs/MetaTool.cc

```diff
@@ -0,0 +1,849 @@
+        for (auto i = io_data_v.begin(); i != io_data_v.end(); ++i)
+            if ((r = _file_meta(op, **i)) == 1)
+                break;
+    }
+    if (r == 1){
+        if (op.ancestors.size() > 0){
+            return 0;
+        }else {
+            std::cerr << "no trace for obj (ino:" << op.top_op()->ino <<")??" << std::endl;
+            return -1;
+        }
+    }else if (op.top_op()->sub_ino_t == meta_op::INO_DIR){
+        std::cerr << "\tmaybe it's a file(ino:" << op.top_op()->ino << ")" << std::endl;
+        op.top_op()->sub_ino_t = meta_op::INO_F;
+        return 1;
+    }
+    
+    std::cerr << "can't get (ino:" << op.top_op()->ino <<")trace??" << std::endl;
+    return -1;
+}
+
+int MetaTool::_file_meta(meta_op &op, librados::IoCtx& io){
+    std::string oid = obj_name(op.top_op()->ino);
+    bufferlist pointer_bl;
+    std::map<std::string, bufferlist> attrset;
+    int r = 0;
+    bool have_data = false;
+    r = io.getxattrs (oid.c_str(), attrset);
+    if (0 == r){
+        std::stringstream ds;
+        std::string format = "json";
+        Formatter* f = Formatter::create(format);
+        auto item = attrset.find("parent");
+        if (item != attrset.end()){
+            inode_backtrace_t i_bt;
+            try {
+                bufferlist::const_iterator q = item->second.cbegin();
+                i_bt.decode(q);
+                f->open_array_section("info");
+                have_data = true;
+                op.ancestors = i_bt.ancestors;
+                f->dump_string("type", "--i_bt--");
+                f->open_object_section("parent");
+                i_bt.dump(f);
+                f->close_section();
+            }catch (buffer::error &e){
+                cerr << "failed to decode parent of " << oid << std::endl;
+                return -1;
+            }
+        }else{
+            cerr << oid << " in " << io.get_pool_name()  << " , but no parent" << std::endl;
+            return -1;
+        }
+        
+        item = attrset.find("layout");
+        if (item != attrset.end()){
+            file_layout_t layout;
+            try {
+                auto q = item->second.cbegin();
+                layout.decode(q);
+                f->dump_string("type", "--layout--");
+                f->open_object_section("layout");
+                layout.dump(f);
+                f->close_section();
+                
+            }catch (buffer::error &e){
+                cerr << "failed to decode layout of " << oid << std::endl;
+                return -1;
+            }
+        }else{
+            cerr << oid << " in " << io.get_pool_name()  << " , but no layout" << std::endl;
+        }
+        if (have_data){
+            f->close_section();
+            f->flush(ds);
+            if (_debug)
+                cout << op.top_op()->ino << " : "<< ds.str() << std::endl;
+            return 1;
+        }
+    }
+    return 0;
+}
+int MetaTool::get_data() {
```

Comment:
what does this function do

Reply:
remove it now. this method is used to access journal of mds.